nupm/publish.nu

```diff
@@ -0,0 +1,101 @@
+            start: $span.start
+            end: $span.end
+        }
+    }
+}
+
+# publish a package
+#
+# > **Warning**  
+# > this command is in a very early stage, it is only meant to generate the
+# > metadata of all the files in the package with the `--generate-metadata`
+# > option.
+export def main [
+    --generate-metadata: bool  # only generate the package metadata file
+    --host: string  # where the package is hosted (used with `--generate-metadata`)
+    --path: string = ""  # the path to the package (used with `--generate-metadata`)
+    --repo: string  # the name of the repo holding the package (used with `--generate-metadata`)
+    --revision: string  # the revision of the repo holding the package (used with `--generate-metadata`)
+] {
+    if $generate_metadata {
+        # TODO: add support for a `.nupmignore` file in the root of a package
+        # to not add files into the package.
+
+        let PACKAGE_FILE = "package.nuon"
+        let METADATA_FILE = "package.files.nuon"
+
+        log info "generating package file metadata file"
+
+        log debug "checking arguments to generate metadata"
+        for option in [
+            [name value]; [host $host] [repo $repo] [revision $revision]
+        ] {
+            if $option.value == null {
+                throw-error $"missing_argument(ansi reset): --($option.name) is required with --generate-metadata"
+            }
+        }
+
+        log debug "building the base download URL"
+        let base_download_url = match ($host | str downcase) {
+            "github.com" | "github" | "gh" => {
+                scheme: "https"
+                host: "raw.githubusercontent.com"
+                path: $"/($repo)/($revision)"
+            },
+            _ => (
+                throw-error
+                    "host_not_supported"
+                    "not a supported host"
+                    --span (metadata $host | get span)
+            ),
+        }
+
+        log debug "listing all files in the package"
+        let package_files = (
+            ls ($path | path join "**" "*") | where type == "file"
+        )
+        log debug $"excluding `($PACKAGE_FILE)` and `($METADATA_FILE)`"
+        let package_files = (
+            $package_files | where {|it|
+                let filename = ($it.name | path basename)
+                ($filename != $PACKAGE_FILE) and ($filename != $METADATA_FILE)
+            }
+        )
+
+        log debug "building metadata for package files:"
+        $package_files | each {|file|
+            log debug $file.name
+            {
+                checksum: ($file.name | open --raw | hash sha256)
+                name: ($file.name | if $path != "." { str replace $path "" } else {} | str trim --left --char "/")
+                raw-url: ($base_download_url | update path { path join $file.name } | url join)
+                supported-os: ($nu.os-info | reject kernel_version)
```

Comment:
Defaulting to current OS is fine but it should be overridable, so I can easily say `it works as long as there's nushell` without playing with the metadatafile manually

Reply:
maybe one could just run a command on `package.files.json` for now?

Reply:
For now, I would just add an optional flag to provide valid OSes and default to what you've written here if the flag was not provided

Reply:
should be done in 82aed49 :wink: